src/main/java/bsh/BSHBlock.java

```diff
@@ -31,101 +31,110 @@
+    * the new namespace is released and the old namespace restored.
+    * <p>
+    * There are some situations where a new name space is not desired, for
+    * example:
+    * <ul>
+    * <li>BshMethod.invokeImpl()
+    * <br>The method namespace has already been set up containing
+    * the formal parameters.
+    * <li>BSHAllocationExpression.constructWithInterfaceBody()
+    * <br>The caller sets up a namespace the same as if it would be
+    * done using overrideChild=false.
+    * <li>BSHEnumConstant.eval()
+    * <br>The enum constants should be set within the current namespace.
+    * <li>BSHTryStatement.eval()
+    * <br>Holds the catch parameter and swaps it on the stack after initializing
+    * </ul>
+    * In these situations where the overrideChild flag has been set to true *no*
+    * new BlockNamespace will be swapped onto the stack and the eval will happen
+    * in the current top namespace.
+    * @param callstack the stack of namespace chains used to resolve names
+    * @param interpreter the interpreter object used for evaluation
+    * @param overrideNamespace whether a new namespace is required
+    * @return the result from evaluating the block
+    * @throws EvalError rolls exceptions back up to the user */
+    public Object eval( CallStack callstack, Interpreter interpreter,
+            boolean overrideNamespace ) throws EvalError {
+
+        if ( isSynchronized ) {
             // First node is the expression on which to sync
             Node exp = jjtGetChild(0);
-            syncValue = exp.eval(callstack, interpreter);
-        }
-
-        Object ret;
-        if ( isSynchronized ) // Do the actual synchronization
-            synchronized( syncValue )
-            {
-                ret = evalBlock(
+            Object syncValue = exp.eval(callstack, interpreter);
+            synchronized( syncValue ) { // Do the actual synchronization
+                return evalBlock(
                     callstack, interpreter, overrideNamespace, null/*filter*/);
             }
-        else
-                ret = evalBlock(
-                    callstack, interpreter, overrideNamespace, null/*filter*/);
-
-        return ret;
+        }
+        return evalBlock(
+                callstack, interpreter, overrideNamespace, null/*filter*/);
     }
 
-    Object evalBlock(
-        CallStack callstack, Interpreter interpreter,
-        boolean overrideNamespace, NodeFilter nodeFilter )
-        throws EvalError
-    {
+    Object evalBlock( CallStack callstack, Interpreter interpreter,
+            boolean overrideNamespace, NodeFilter nodeFilter ) throws EvalError {
+
         Object ret = Primitive.VOID;
         NameSpace enclosingNameSpace = null;
         if ( !overrideNamespace )
             enclosingNameSpace = callstack.swap(
-                    blockspaces.get(callstack.top()));
+                    BlockNameSpace.getInstance(callstack.top(), blockId));
```

Comment:
>  I still cannot see the point for using a map for keeping track of cached namespaces. Isn't one recycled namespace just as good as another, as long as they are truly free and have no references?

I was also starting to think on the lines of a pool of reusable namespaces. But there is always room for improvement. However, while there are still unresolved issues, I chose to use a new namespace for every unique block, just like we would without the cache, this means that most blocks are never reused as they are only ever executed once, and accrues an extra penalty with no added benefit. This is outweighed by the benefits achieved for the blocks that are reused, while at the same time giving us a solid foundation to work from.

Reply:
>  if the `this` variable escapes.

I don't understand what you mean by this and the referenced post did not shed any more light either. According to me `this` issues were resolved in this patch.

The new implementation only reuses namespace when the block is looped, iterated, recursed, or in any other way called more than once. This was specifically done so that we can root cache out as being the culprit.